pkg/controller/storagecluster/reconcile.go

```diff
@@ -409,3 +456,63 @@ func newCephDaemonResources(custom map[string]corev1.ResourceRequirements) map[s
 
 	return resources
 }
+
+// newStorageClassDeviceSets converts a list of StorageDeviceSets into a list of Rook StorageClassDeviceSets
+func newStorageClassDeviceSets(storageDeviceSets []ocsv1.StorageDeviceSet, topologyMap ocsv1.NodeTopologyMap) []rook.StorageClassDeviceSet {
+	var storageClassDeviceSets []rook.StorageClassDeviceSet
+
+	for _, ds := range storageDeviceSets {
+		resources := ds.Resources
+		if resources.Requests == nil && resources.Limits == nil {
+			resources = defaults.DaemonResources["osd"]
+		}
+
+		topologyKey := ds.TopologyKey
+		topologyKeyValues := []string{}
+		noPlacement := ds.Placement.NodeAffinity == nil && ds.Placement.PodAffinity == nil && ds.Placement.PodAntiAffinity == nil
+
+		if topologyKey != "" && noPlacement {
+			topologyKey, topologyKeyValues = topologyMap.GetKeyValues(topologyKey)
+		}
+
+		for i := 0; i <= ds.Replica; i++ {
+			placement := ds.Placement
+			portable := ds.Portable
+
+			if noPlacement {
+				placement = defaults.DaemonPlacements["osd"]
+				if len(topologyKeyValues) >= 3 {
+					portable = true
+					placement.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution[0].PodAffinityTerm.TopologyKey = topologyKey
+
+					topologyIndex := i
+					for topologyIndex >= len(topologyKeyValues) {
```

Comment:
Would it make more sense to use the modulus here to wrap around the index when there are more ds than topologykeyvalues?

Reply:
Not a bad idea, I completely forgot about it!